### DISCUSSION TITLE
思考一下二分查找

### DISCUSSION THREAD

**Discussion by binderclip** [binderclip/clip#35] [notes]
## 跳出代码实现来想一下

目标：有一组元素是有序的，要查找其中的一个

1.  看这一组中间的那一个的值 m（奇数取正中间那个，偶数取中线左边那个）
2.  如果是要查找的值那么找到了目标，结束
3.  如果 m 大于目标值，那么目标可能在左边那些里面，m 小于目标值则在右边那些元素里面
4.  重复上面的循环，直到找到结束，或者找到了最后一个元素也没找到

## 回归到代码上

```go
func binarySearch(nums []int, target int) (index int) {
    if len(nums) == 0 {
        return -1
    }
    il, ir := 0, len(nums) - 1 // 确定初始范围
    for il <= ir {
        m := (il + ir) / 2 // 利用整数整除的特性，就可以实现“奇数取正中间那个，偶数取中线左边那个”
        t := nums[m]
        if t == target {
            return m // 如果发现了会直接返回
        } else if t < target {
            il = m + 1 // 没找到，但可以进一步缩小范围
        } else {
            ir = m - 1 // 没找到，但可以进一步缩小范围
        }
    }
    // 退出循环说明找完了，但由于上面的循环中如果定位到了会直接返回，所以这里的结束也能表示出没找到
    return -1
}
```

## 扩展一下，查找第一个等于的

先做一下抽象，假设列表是盖着的扑克牌，虽然是有序的但是没翻开的部分不清楚具体值，翻开中间的那个牌（查看中间的元素值）之后有三种情况：

```
X X X X X
    <     翻开中间，发现比目标小，则查找右边的
      X X
```

```
X X X X X
    =     翻开中间，发现等于目标，但是有可能左边还有等于目标的，所以要继续往左边找
X X
```

```
X X X X X
    >     翻开中间，发现大于目标，要继续往左边找
X X
```

这次的查找不像最开始的那种找到了就可以提前退出，会一直查找到最后一个元素，结束之后回来看最后一个元素有三种可能：

```
X
= 等于，因为之前发现等于的之后只会继续往左边查，所以最后发现是等于一定是等于的里面最靠左的
```

```
X
> 大于，因为之前发现等于的之后只会继续往左边查，所以最后发现是大于那全部的元素肯定都大于目标值
```

```
X
< 小于，如果它是列表的最后一个元素那么说明全部元素都小于目标值，如果不是最后一个元素那么它右边的元素一定是 >= 目标值的，需要另外检查一下
```

或者在循环的过程中如果发现等于就进行一次 index 的记录，等结束的时候也就是需要的值。

## 其他场景

*   查找最后一个等于的，可以用类似的逻辑处理
*   查找最后一个小于的
*   查找第一个大于的
*   ……

## Go 标准库里面的 search

函数的结果是要查找的值，或者要查找的值应该插入的位置；  
实际如果是要查找的话，基于函数返回的值还是要继续检查一下 index 是否超出范围以及 index 对应结果是否等于目标值；

```go
// SearchInts searches for x in a sorted slice of ints and returns the index
// as specified by Search. The return value is the index to insert x if x is
// not present (it could be len(a)).
// The slice must be sorted in ascending order.
//
func SearchInts(a []int, x int) int {
    return Search(len(a), func(i int) bool { return a[i] >= x })
}

func Search(n int, f func(int) bool) int {
    // Define f(-1) == false and f(n) == true.
    // Invariant: f(i-1) == false, f(j) == true.
    i, j := 0, n
    for i < j {
        h := int(uint(i+j) >> 1) // avoid overflow when computing h
        // i ≤ h < j
        if !f(h) {
            i = h + 1 // preserves f(i-1) == false
        } else {
            j = h // preserves f(j) == true
        }
    }
    // i == j, f(i-1) == false, and f(j) (= f(i)) == true  =>  answer is i.
    return i
}
```

## 关联题目

- [34. 在排序数组中查找元素的第一个和最后一个位置 - 力扣（LeetCode）](https://leetcode-cn.com/problems/find-first-and-last-position-of-element-in-sorted-array/)

**Comment by ilumer** [comment]
![image](https://user-images.githubusercontent.com/10941253/151648841-86338fcf-714e-43db-8d0d-0e81333b3b3f.png)
这里可能会出现溢出的情况
```
m := il+(ir-il) / 2
```

**Comment by binderclip** [comment]
> ![image](https://user-images.githubusercontent.com/10941253/151648841-86338fcf-714e-43db-8d0d-0e81333b3b3f.png)
> 
> 这里可能会出现溢出的情况
> 
> ```
> m := il+(ir-il) / 2
> ```

@ilumer  确实是的！当时也注意了标准库里面的 \`h := int(uint(i+j) >> 1)\`
